### PR TITLE
Intercom - New Actions

### PR DESCRIPTION
**Pull request by michelle0927**
New actions:
**Create a Note** | Creates a note for a specific user
**Send Incoming Message** | Send a message from a user into your Intercom app

Addresses issue: https://github.com/PipedreamHQ/pipedream/issues/1499

### CODE REVIEW

**Comment by dylburger** [comment]
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/intercom/actions/create-note/create-note.mjs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    },
+  },
+  async run({ $ }) {
+    const {
+      userId,
+      body,
+    } = this;
+    const { id: adminId } = await this.intercom.getAdmin($);
+    const res = await this.intercom.createNote(userId, adminId, body, $);
+    $.export("summary", "Successfully created note");
```
Comment (jcortes) [inline components/intercom/actions/create-note/create-note.mjs:34]: Optionally I guess you can show the note id here
```suggestion
    const res = await this.intercom.createNote(userId, adminId, body, $);
    $.export("summary", `Successfully created note with ID ${res.id}`);
```

File: components/intercom/actions/send-incoming-message/send-incoming-message.mjs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    const { role } = await this.intercom.getContact(userId, $);
+    const data = {
+      from: {
+        type: role,
+        id: userId,
+      },
+      body,
+    };
+    const res = await this.intercom.createConversation(data, $);
+    $.export("summary", "Message sent successfully");
```
Comment (jcortes) [inline components/intercom/actions/send-incoming-message/send-incoming-message.mjs:42]:
```suggestion
    const res = await this.intercom.createConversation(data, $);
    $.export("summary", `Message sent successfully with ID ${res.id}`);
```

File: components/intercom/intercom.app.mjs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        url,
+        endpoint,
+        data,
+        $,
+      } = opts;
+      const config = {
+        method,
+        url: url
+          ? url
+          : `https://api.intercom.io/${endpoint}`,
```
Comment (jcortes) [inline components/intercom/intercom.app.mjs:60]:
```suggestion
        url: url ?? `https://api.intercom.io/${endpoint}`,
```

File: components/intercom/intercom.app.mjs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        headers: {
+          Authorization: `Bearer ${this.$auth.oauth_access_token}`,
+          Accept: "application/json",
+        },
+        data,
+      };
+      return await axios($ || this, config);
```
Comment (jcortes) [inline components/intercom/intercom.app.mjs:67]: Optionally I would remove this `await` and others in `return` statement.

File: components/intercom/intercom.app.mjs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      let results = null;
+      let done = false;
+      let items = [];
+      while ((!results || results?.pages?.next) && !done) {
+        const startingAfter = results?.pages?.next?.starting_after || null;
+        const endpoint = `${itemType}/${isSearch
+          ? "search"
+          : ""}${startingAfter
+          ? "?starting_after=" + startingAfter
+          : ""}`;
```
Comment (jcortes) [inline components/intercom/intercom.app.mjs:91]: Optionally. I think these lines are a little hard to understand, Maybe I would recommend doing it like this?
```suggestion
        const search = isSearch && "/search" || "";
        const startingAfterParam = startingAfter && `?starting_after=${startingAfter}` || "";
        const endpoint = `${itemType}${search}${startingAfterParam}`;
```

File: components/intercom/package.json [hunk 6]
```diff
@@ -1,18 +1,15 @@
...
+    "main": "intercom.app.mjs",
     "keywords": [
         "pipedream",
         "intercom"
     ],
     "homepage": "https://pipedream.com/apps/intercom",
     "author": "Pipedream <support@pipedream.com> (https://pipedream.com/)",
     "license": "MIT",
-    "dependencies": {
-        "axios": "^0.21.1"
```
Comment (jcortes) [inline components/intercom/package.json:14]: I think you should uninstall `axios` and then `npm i @pipedream/platform`

**Review by jcortes — CHANGES_REQUESTED** [review]
Looks good to me. Just a few comments (optional suggestions).

**Review by jcortes — APPROVED** [review]
It looks good to me, thanks!

**Comment by dannyroosevelt** [comment]
@michelle0927 plz merge when ready!

**Pull request merged**